### PR TITLE
fix: prelu perf gap on Unet

### LINKED ISSUE #3715: 🐛 [Bug] Torch-TRT's prelu is slower than ONNX-TRT
##  Bug Description

In Monai/Unet, I observed that Torch-TRT's prelu is slower than ONNX-TRT, and what is worse is that using prelu converter is even slower than decomposing it. The reason seems to be additional layers are inserted into the graph:

```
ONNX-TRT:
{ "name" : "PWN(val_9, PWN(node_prelu))", "timeMs" : 1.10378, "averageMs" : 0.00951531, "medianMs" : 0.009216, "percentage" : 0.0934733 }

Torch-TRT w/o decomposition:
{ "name" : "__myl_Cast_myl3_1", "timeMs" : 0.95728, "averageMs" : 0.00878239, "medianMs" : 0.009216, "percentage" : 0.0743985 }
{ "name" : "model.0.conv.unit0.adn.A/_prelu_kernel_constant_1", "timeMs" : 0, "averageMs" : 0, "medianMs" : 0, "percentage" : 0 }
{ "name" : "PWN([PARAMETRIC_RELU]-[aten_ops._prelu_kernel.default]-[model.0.conv.unit0.adn.A/_prelu_kernel])", "timeMs" : 2.87757, "averageMs" : 0.0263997, "medianMs" : 0.026624, "percentage" : 0.223641 }

Torch-TRT w/ decomposition:
{ "name" : "__myl_MulGtrSele_myl3_1", "timeMs" : 0.946529, "averageMs" : 0.00884606, "medianMs" : 0.009216, "percentage" : 0.0707629 }
```

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

In Monai/Unet, I observed that Torch-TRT's prelu is slower than ONNX-TRT, and what is worse is that using prelu converter is even slower than decomposing it.

Update: The perf issue is because `to_copy` forces to insert Cast Layer.

Fixes #3715 #3723 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by narendasan — APPROVED** [review]
LGTM

**Comment by narendasan** [comment]
If possible consider adding a test, we talked a bit about a new INetwork verification test based on expected numbers of ops. Might be a simple case to build the harness for

**Comment by zewenli98** [comment]
@narendasan I'll merge this PR now and cherrypick to the release 2.8, and then work on the number of ops check in another PR for the next release. sounds like a plan?

**Comment by zewenli98** [comment]
there's an error. will merge after fixing

File: py/torch_tensorrt/dynamo/conversion/impl/prelu.py [hunk 1]
```diff
@@ -15,6 +16,12 @@ def prelu(
     input: TRTTensor,
     weight: TRTTensor,
 ) -> TRTTensor:
-    layer = ctx.net.add_parametric_relu(input, weight)
+    # TRT requires that the slopes tensor must be unidirectional broadcastable to the input tensor:
+    # the rank of the two tensors must be the same, and all dimensions of the slopes tensor must
+    # either equal the input tensor or be 1. The output tensor has the same shape as the input tensor.
+    input, weight = impl.elementwise.broadcast(
+        ctx, input, weight, f"{name}_broadcast_input", f"{name}_broadcast_weight"
+    )
```
Comment (HolyWu) [inline py/torch_tensorrt/dynamo/conversion/impl/prelu.py]: This is probably redundant because PyTorch already reshapes the weight beforehand so we already receive the weight of the required shape.


```py
import torch
import torch_tensorrt


class MyModule(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()

    def forward(self, x: torch.Tensor, weight: torch.Tensor) -> torch.Tensor:
        return torch.nn.functional.prelu(x, weight)


with torch.inference_mode():
    model = MyModule().eval().cuda().half()

    inputs = (
        torch.randn(1, 3, 224, 224, dtype=torch.half, device="cuda"),
        torch.randn(3, dtype=torch.half, device="cuda"),
    )

    exported_program = torch.export.export(model, inputs)

    with torch_tensorrt.dynamo.Debugger():
        trt_model = torch_tensorrt.dynamo.compile(
            exported_program, inputs, enabled_precisions={torch.half}, min_block_size=1
        )

    torch.testing.assert_close(trt_model(*inputs), model(*inputs), rtol=5e-3, atol=5e-3)
```

```py
10:39:52 - DEBUG - Invoking DynamoPassManager and applying lowering passes: [<function remove_detach at 0x000001A0385E4EA0>]
10:39:52 - DEBUG - Removed 0 detach nodes:
graph():
    %x : [num_users=1] = placeholder[target=x]
    %weight : [num_users=1] = placeholder[target=weight]
    %prelu : [num_users=1] = call_function[target=torch.ops.aten.prelu.default](args = (%x, %weight), kwargs = {})
    return (prelu,)
10:39:52 - DEBUG - Input graph: graph():
    %x : [num_users=1] = placeholder[target=x]
    %weight : [num_users=1] = placeholder[target=weight]
    %_reshape_copy : [num_users=1] = call_function[target=torch.ops.aten._reshape_copy.default](args = (%weight, [1, 3, 1, 1]), kwargs = {})
    %_prelu_kernel : [num_users=1] = call_function[target=torch.ops.aten._prelu_kernel.default](args = (%x, %_reshape_copy), kwargs = {})
    return (_prelu_kernel,)
```

**Comment by zewenli98** [comment]
```
def forward(self, x):
    x_1 = x + 1
    y = torch.ops.aten._to_copy.default(x_1, dtype=torch.float)
    z = torch.ops.aten._to_copy.default(x_1, dtype=torch.float)
    return y, z
```
this case fails because `self.engine.num_io_tensors=2`, `len(self.input_names)=1`, `len(self.output_names)=2`, which doesn't meet the assertion:
```
assert self.engine.num_io_tensors == len(self.input_names) + len(self.output_names)
```

self.input_names: ['x']
self.output_names: ['output0', 'output1']

Since `output0` is invalid, it gets `self.engine.num_io_tensors=2`.

File: py/torch_tensorrt/dynamo/conversion/impl/prelu.py (hunk 1 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/prelu.py]: in case that _prelu_kernel is directly called?

Comment (HolyWu) [reply py/torch_tensorrt/dynamo/conversion/impl/prelu.py]: then you still get the wrong shape of weight because `broadcast` prepend 1s to the beginning and hence you get `(1, 1, 1, 3)` rather than `(1, 3, 1, 1)` with the above example.

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/prelu.py]: thanks for the bug report. I'll revert and leave pytorch to do the reshape

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 2]
```diff
@@ -941,7 +941,14 @@ def output(self, target: str, args: Any, kwargs: Any) -> List[Any]:
                 f"Specified output dtypes ({len(self.output_dtypes)}) differ from number of outputs ({len(outputs)})"
             )
 
+        marked_outputs_ids = []
         for i, output in enumerate(outputs):
+            # In some cases, the same output tensor may be marked multiple times, such as _to_oppy,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Small typo

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 3]
```diff
@@ -941,7 +941,14 @@ def output(self, target: str, args: Any, kwargs: Any) -> List[Any]:
...
+            # so we skip marking if the output is already marked
+            if id(output) in marked_outputs_ids:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:948]: Where does this id function come from?

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:948]: `id()` is a built-in function that returns the "identity" of an object.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 4]
```diff
@@ -1094,7 +1094,7 @@ def aten_ops_clone_copy_dtype(
         name,
         args[0],
         kwargs.get("dtype", args[0].dtype),
-        force_layer=True,
+        force_layer=False,
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: I think there might be cases where we might need to actually force_layer=True. Do you know when that would be useful?. Also consider adding a comment here conveying that force_layer=False results in better performance.

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: Do you know when we need force_layer=True? My understanding is that 1) Since `cast_trt_tensor` would be called even if the force_layer=False, it only adds Cast Layer if necessary. 2) If an operator satisfies `not is_only_operator_on_placeholder`, that means there's other operator on the placeholder or the lowering pass `repair_input_as_output` to promise "input is not output", so we don't have to explicitly add Cast Layer.
